unittests/finality_test_cluster.hpp

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <eosio/chain/hotstuff/finalizer_authority.hpp>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsign-compare"
+#include <boost/test/unit_test.hpp>
+#pragma GCC diagnostic pop
+#include <eosio/testing/tester.hpp>
+
+// Set up a test network which consists of 3 nodes:
+//   * node0 produces blocks and pushes them to node1 and node2;
+//     node0 votes the blocks it produces internally.
+//   * node1 votes on the proposal sent by node0
+//   * node2 votes on the proposal sent by node0
+// Each node has one finalizer: node0 -- "node0"_n, node1 -- "node1"_n, node2 -- "node2"_n.
+// Quorum is set to 2.
+// After starup up, IF are activated on both nodes.
+//
+// APIs are provided to modify/delay/reoder/remove votes from node1 and node2 to node0.
+
+
+class finality_test_cluster {
+public:
+
+   enum class vote_mode {
+      strong,
+      weak,
+   };
+
+   // Construct a test network and activate IF.
+   finality_test_cluster();
+
+   // node0 produces a block and pushes it to node1 and node2
+   void produce_and_push_block();
+
+   // send node1's vote identified by "index" in the collected votes
+   eosio::chain::vote_status process_node1_vote(uint32_t vote_index, vote_mode mode = vote_mode::strong);
+
+   // send node1's latest vote
+   eosio::chain::vote_status process_node1_vote(vote_mode mode = vote_mode::strong);
+
+   // send node2's vote identified by "index" in the collected votes
+   eosio::chain::vote_status process_node2_vote(uint32_t vote_index, vote_mode mode = vote_mode::strong);
+
+   // send node2's latest vote
+   eosio::chain::vote_status process_node2_vote(vote_mode mode = vote_mode::strong);
```

Comment:
Why not a single finction process_node_vote taking a node parameter?

Reply:
Those are provided as a convenient method for the users and hide the implementation details from them. This way the actual test cases do not need t be changed even if the implementation of test cluster changes again.